File: src/ios/BabyCryAnalyzer/Audio/AudioCaptureModule.m
//
// AudioCaptureModule.m
// BabyCryAnalyzer
//
// React Native bridge module for iOS audio capture with enhanced error handling
// and performance monitoring capabilities.
//
// Foundation: iOS 14.0+
// React: 0.71+
//

#import "AudioCaptureModule.h"
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>
#import <AVFoundation/AVFoundation.h>

// MARK: - Constants
static NSString *const kEventAudioLevelUpdate = @"onAudioLevelUpdate";
static NSString *const kEventCryDetected = @"onCryDetected";
static NSString *const kEventError = @"onError";
static NSString *const kEventMetricsUpdate = @"onMetricsUpdate";

@interface AudioCaptureModule () <AudioProcessorDelegate>

@property (nonatomic, strong) AudioProcessor *audioProcessor;
@property (nonatomic, strong) AudioConfiguration *configuration;
@property (nonatomic, assign) BOOL hasListeners;
@property (nonatomic, strong) dispatch_queue_t processingQueue;
@property (nonatomic, strong) NSMutableDictionary *performanceMetrics;
@property (nonatomic, strong) NSLock *stateLock;

@end

@implementation AudioCaptureModule

// MARK: - RCTBridgeModule Implementation
RCT_EXPORT_MODULE();

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[kEventAudioLevelUpdate,
             kEventCryDetected,
             kEventError,
             kEventMetricsUpdate];
}

// MARK: - Lifecycle Methods
- (instancetype)init {
    if (self = [super init]) {
        _processingQueue = dispatch_queue_create("com.babycryanalyzer.capture", DISPATCH_QUEUE_SERIAL);
        _performanceMetrics = [NSMutableDictionary dictionary];
        _stateLock = [[NSLock alloc] init];
        
        // Initialize audio configuration
        _configuration = [[AudioConfiguration alloc] initWithSampleRate:44100.0
                                                          bufferSize:4096
                                                     noiseThreshold:-50.0
                                              cryDetectionThreshold:0.85
                                               minimumCryDuration:1.5];
        
        // Initialize audio processor
        NSError *error;
        _audioProcessor = [[AudioProcessor alloc] initWithConfig:_configuration error:&error];
        if (error) {
            NSLog(@"Failed to initialize AudioProcessor: %@", error);
        }
        _audioProcessor.delegate = self;
    }
    return self;
}

// MARK: - React Native Methods
RCT_EXPORT_METHOD(startRecording:(NSDictionary *)options
                  withResolver:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject) {
    dispatch_async(_processingQueue, ^{
        [self.stateLock lock];
        @try {
            if (!self.audioProcessor) {
                reject(@"INIT_ERROR", @"Audio processor not initialized", nil);
                return;
            }
            
            NSError *error;
            BOOL success = [self.audioProcessor startRecording:&error];
            
            if (success) {
                [self startPerformanceMonitoring];
                resolve(@{@"status": @"recording"});
            } else {
                reject(@"RECORDING_ERROR",
                       error.localizedDescription,
                       error);
            }
        } @finally {
            [self.stateLock unlock];
        }
    });
}

RCT_EXPORT_METHOD(stopRecording:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject) {
    dispatch_async(_processingQueue, ^{
        [self.stateLock lock];
        @try {
            [self.audioProcessor stopRecording];
            [self stopPerformanceMonitoring];
            
            resolve(@{
                @"status": @"stopped",
                @"metrics": self.performanceMetrics
            });
        } @finally {
            [self.stateLock unlock];
        }
    });
}

RCT_EXPORT_METHOD(isRecording:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject) {
    resolve(@{@"recording": @(self.audioProcessor.isRecording)});
}

RCT_EXPORT_METHOD(getMetrics:(RCTPromiseResolveBlock)resolve
                  withRejecter:(RCTPromiseRejectBlock)reject) {
    resolve(self.performanceMetrics);
}

// MARK: - RCTEventEmitter Methods
- (void)startObserving {
    [self.stateLock lock];
    self.hasListeners = YES;
    [self.stateLock unlock];
}

- (void)stopObserving {
    [self.stateLock lock];
    self.hasListeners = NO;
    [self.stateLock unlock];
}

// MARK: - AudioProcessorDelegate Methods
- (void)audioProcessorDidStartRecording {
    if (!self.hasListeners) return;
    [self sendEventWithName:kEventMetricsUpdate
                      body:@{@"status": @"recording"}];
}

- (void)audioProcessorDidStopRecording {
    if (!self.hasListeners) return;
    [self sendEventWithName:kEventMetricsUpdate
                      body:@{@"status": @"stopped"}];
}

- (void)audioProcessorDidDetectAudioLevel:(float)audioLevel {
    if (!self.hasListeners) return;
    [self sendEventWithName:kEventAudioLevelUpdate
                      body:@{@"level": @(audioLevel)}];
}

- (void)audioProcessorDidDetectCryWithPatternType:(NSString *)patternType
                                      confidence:(float)confidence {
    if (!self.hasListeners) return;
    [self sendEventWithName:kEventCryDetected
                      body:@{
                          @"type": patternType,
                          @"confidence": @(confidence),
                          @"timestamp": @([[NSDate date] timeIntervalSince1970])
                      }];
}

// MARK: - Private Methods
- (void)startPerformanceMonitoring {
    self.performanceMetrics[@"startTime"] = @([[NSDate date] timeIntervalSince1970]);
    self.performanceMetrics[@"bufferOverruns"] = @(0);
    self.performanceMetrics[@"processingTime"] = @(0);
}

- (void)stopPerformanceMonitoring {
    NSTimeInterval endTime = [[NSDate date] timeIntervalSince1970];
    NSTimeInterval startTime = [self.performanceMetrics[@"startTime"] doubleValue];
    self.performanceMetrics[@"duration"] = @(endTime - startTime);
    self.performanceMetrics[@"completed"] = @YES;
}

@end

File: src/ios/BabyCryAnalyzer/Audio/AudioCaptureModule.swift
//
// AudioCaptureModule.swift
// BabyCryAnalyzer
//
// React Native bridge module for iOS audio capture with enhanced error handling
// and performance monitoring capabilities.
//
// Foundation: iOS 14.0+
// React: 0.71+
//

import Foundation
import React

// MARK: - Constants
private let kModuleName = "AudioCaptureModule"
private let kEventAudioLevelUpdate = "onAudioLevelUpdate"
private let kEventCryDetected = "onCryDetected"
private let kEventError = "onError"
private let kEventMetricsUpdate = "onMetricsUpdate"

// MARK: - AudioCaptureModule
@objc(AudioCaptureModule)
class AudioCaptureModule: RCTEventEmitter {
    
    // MARK: - Private Properties
    private var audioProcessor: AudioProcessor?
    private var configuration: AudioConfiguration
    private var hasListeners: Bool = false
    private let processingQueue: DispatchQueue
    private let stateLock: NSLock
    private var metrics: ProcessingMetrics
    
    // MARK: - Initialization
    override init() {
        self.configuration = AudioConfiguration(
            sampleRate: 44100.0,
            bufferSize: 4096,
            noiseThreshold: -50.0,
            cryDetectionThreshold: 0.85
        )
        self.processingQueue = DispatchQueue(label: "com.babycryanalyzer.capture", qos: .userInitiated)
        self.stateLock = NSLock()
        self.metrics = ProcessingMetrics()
        
        super.init()
        
        // Initialize audio processor with configuration
        do {
            self.audioProcessor = try AudioProcessor(config: configuration)
            self.audioProcessor?.delegate = self
        } catch {
            print("Failed to initialize AudioProcessor: \(error)")
        }
    }
    
    // MARK: - RCTEventEmitter Override
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override func supportedEvents() -> [String] {
        return [
            kEventAudioLevelUpdate,
            kEventCryDetected,
            kEventError,
            kEventMetricsUpdate
        ]
    }
    
    override func startObserving() {
        stateLock.lock()
        hasListeners = true
        stateLock.unlock()
    }
    
    override func stopObserving() {
        stateLock.lock()
        hasListeners = false
        stateLock.unlock()
    }
    
    // MARK: - Public Methods
    @objc(startRecording:withRejecter:)
    func startRecording(_ resolve: @escaping RCTPromiseResolveBlock,
                       withRejecter reject: @escaping RCTPromiseRejectBlock) {
        processingQueue.async { [weak self] in
            guard let self = self else {
                reject("ERROR", "Module deallocated", nil)
                return
            }
            
            guard let audioProcessor = self.audioProcessor else {
                reject("INIT_ERROR", "Audio processor not initialized", nil)
                return
            }
            
            switch audioProcessor.startRecording() {
            case .success:
                self.metrics.startMonitoring()
                resolve(["status": "recording"])
                
            case .failure(let error):
                self.sendEvent(withName: kEventError, body: [
                    "code": "RECORDING_ERROR",
                    "message": error.localizedDescription
                ])
                reject("RECORDING_ERROR", error.localizedDescription, error)
            }
        }
    }
    
    @objc(stopRecording:withRejecter:)
    func stopRecording(_ resolve: @escaping RCTPromiseResolveBlock,
                      withRejecter reject: @escaping RCTPromiseRejectBlock) {
        processingQueue.async { [weak self] in
            guard let self = self else {
                reject("ERROR", "Module deallocated", nil)
                return
            }
            
            self.stateLock.lock()
            defer { self.stateLock.unlock() }
            
            self.audioProcessor?.stopRecording()
            self.metrics.stopMonitoring()
            
            resolve(["status": "stopped", "metrics": self.metrics.summary])
        }
    }
    
    @objc(isRecording:withRejecter:)
    func isRecording(_ resolve: @escaping RCTPromiseResolveBlock,
                    withRejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(["recording": audioProcessor?.isRecording ?? false])
    }
    
    @objc(getMetrics:withRejecter:)
    func getMetrics(_ resolve: @escaping RCTPromiseResolveBlock,
                   withRejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(metrics.current)
    }
}

// MARK: - AudioProcessorDelegate
extension AudioCaptureModule: AudioProcessorDelegate {
    func audioProcessorDidStartRecording() {
        guard hasListeners else { return }
        sendEvent(withName: kEventMetricsUpdate, body: ["status": "recording"])
    }
    
    func audioProcessorDidStopRecording() {
        guard hasListeners else { return }
        sendEvent(withName: kEventMetricsUpdate, body: ["status": "stopped"])
    }
    
    func audioProcessor(didDetectAudioLevel audioLevel: Float) {
        guard hasListeners else { return }
        sendEvent(withName: kEventAudioLevelUpdate, body: ["level": audioLevel])
    }
    
    func audioProcessorDidDetectCry(patternType: String, confidence: Float) {
        guard hasListeners else { return }
        sendEvent(withName: kEventCryDetected, body: [
            "type": patternType,
            "confidence": confidence,
            "timestamp": Date().timeIntervalSince1970
        ])
    }
}

// MARK: - ProcessingMetrics
private struct ProcessingMetrics {
    private var startTime: Date?
    private var processingStats: [String: Any] = [:]
    
    mutating func startMonitoring() {
        startTime = Date()
        processingStats = [:]
    }
    
    mutating func stopMonitoring() {
        startTime = nil
    }
    
    var current: [String: Any] {
        var metrics: [String: Any] = processingStats
        if let start = startTime {
            metrics["duration"] = Date().timeIntervalSince(start)
        }
        return metrics
    }
    
    var summary: [String: Any] {
        var summary = current
        summary["completed"] = true
        return summary
    }
}